### PR TITLE
Create Castep package

### PR DESCRIPTION
**Pull request by tomgreen66**
Created first attempt at Castep package.  Wasn't sure whether to try and build a makefile from first principle or use the supplied makefiles.  Decided to use existing Makefiles system with minor edits to work with Spack.

### CODE REVIEW

File: var/spack/repos/builtin/packages/castep/package.py [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+    properties of materials from first principles.
+    Using density functional theory, it can simulate
+    a wide range of properties of materials
+    proprieties including energetics, structure at
+    the atomic level, vibrational properties,
+    electronic response properties etc.
+    """
+
+    homepage = "https://castep.org"
+    url = "file://{0}/CASTEP-21.11.tar.gz".format(os.getcwd())
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/castep/package.py]: You should be able to use an f-string here (https://docs.python.org/3/tutorial/inputoutput.html).

File: var/spack/repos/builtin/packages/castep/package.py [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+    CASTEP is a leading code for calculating the
+    properties of materials from first principles.
+    Using density functional theory, it can simulate
+    a wide range of properties of materials
+    proprieties including energetics, structure at
+    the atomic level, vibrational properties,
+    electronic response properties etc.
+    """
+
+    homepage = "https://castep.org"
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/castep/package.py]: It doesn't look like the homepage is accessible through the `https` protocol (only `http`).

**Review by tldahlgren — CHANGES_REQUESTED** [review]
This appears to be licensed software (per http://www.castep.org/CASTEP/GettingCASTEP).  I am unable to confirm the versions or version sha256 but there is the homepage change and you may need to add licensing information.

You may want to check out our docs at https://spack.readthedocs.io/en/latest/packaging_guide.html#licensed-software.

**Comment by tomgreen66** [comment]
> This appears to be licensed software (per http://www.castep.org/CASTEP/GettingCASTEP). I am unable to confirm the versions or version sha256 but there is the homepage change and you may need to add licensing information.
> 
> You may want to check out our docs at https://spack.readthedocs.io/en/latest/packaging_guide.html#licensed-software.

It is licenced software but similar to Vasp it doesn't use any runtime licence file to limit access (just a manual download) so nothing further can really be added from my understanding reading that page.

File: var/spack/repos/builtin/packages/castep/package.py (hunk 2 above)
Comment (tomgreen66) [reply var/spack/repos/builtin/packages/castep/package.py]: Sorry about that - fixed.

File: var/spack/repos/builtin/packages/castep/package.py (hunk 1 above)
Comment (tomgreen66) [reply var/spack/repos/builtin/packages/castep/package.py]: Just followed practice from another package - have used f-string here if that is preferred.

**Review by tldahlgren — DISMISSED** [review]

**Comment by tldahlgren** [comment]
Looks like a few style checks need to be fixed:

```
==> Running black checks
--- var/spack/repos/builtin/packages/castep/package.py	2023-12-10 21:40:37.063318+00:00
+++ var/spack/repos/builtin/packages/castep/package.py	2023-12-10 21:40:50.308497+00:00
@@ -22,11 +22,13 @@
     homepage = "http://castep.org/"
     url = f"file://{os.getcwd()}/CASTEP-21.11.tar.gz"
     manual_download = True
 
     version("21.11", sha256="d909936a51dd3dff7a0847c2597175b05c8d0018d5afe416737499408914728f")
-    version("19.1.1.rc2", sha256="1fce21dc604774e11b5194d5f30df8a0510afddc16daf3f8b9bbb3f62748f86a")
+    version(
+        "19.1.1.rc2", sha256="1fce21dc604774e11b5194d5f30df8a0510afddc16daf3f8b9bbb3f62748f86a"
+    )
 
     variant("mpi", default=True, description="Enable MPI build")
     depends_on("rsync", type="build")
     depends_on("blas")
     depends_on("lapack")
@@ -34,11 +36,10 @@
     depends_on("mpi", type=("build", "link", "run"), when="+mpi")
 
     parallel = True
 
     def edit(self, spec, prefix):
-
         if spec.satisfies("%gcc"):
             dlmakefile = FileFilter("LibSource/dl_mg-2.0.3/platforms/castep.inc")
             dlmakefile.filter(r"MPIFLAGS = -DMPI", "MPIFLAGS = -fallow-argument-mismatch -DMPI")
             if self.spec.satisfies("@20:"):
                 platfile = FileFilter("obj/platforms/linux_x86_64_gfortran.mk")
@@ -85,12 +86,10 @@
                     if self.spec.satisfies("@20:"):
                         targetlist.append("ARCH=linux_x86_64_ifort")
                     else:
                         targetlist.append("ARCH=linux_x86_64_ifort19")
 
-
-
         return targetlist
 
     def install(self, spec, prefix):
         mkdirp(prefix.bin)
         make("install", "install-tools", *self.build_targets, "INSTALL_DIR={0}".format(prefix.bin))
would reformat var/spack/repos/builtin/packages/castep/package.py
Oh no! 💥 💔 💥
1 file would be reformatted.
  black found errors
==> Running flake8 checks
var/spack/repos/builtin/packages/castep/package.py:92: [E303] too many blank lines (3)
  flake8 found errors
```

**Comment by haampie** [comment]
@spackbot fix style

**Comment by tomgreen66** [comment]
Apologies - got caught up in other things but have just ran the style fix myself on the branch and hopefully should pass now.

File: var/spack/repos/builtin/packages/castep/package.py [hunk 3]
```diff
@@ -0,0 +1,95 @@
+# Copyright 2013-2023 Lawrence Livermore National Security, LLC and other
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/castep/package.py:1]:
```suggestion
# Copyright 2013-2024 Lawrence Livermore National Security, LLC and other
```

Comment (tldahlgren) [reply var/spack/repos/builtin/packages/castep/package.py:1]: FYI.  This change isn't technically required by the checks this year.

**Review by tldahlgren — APPROVED** [review]
It's been a long time (apologies) so I'm going to merge as is.

**Pull request merged**